File: crates/play-gui/curl-helper/scripts/generate_icon.py
#!/usr/bin/env python3
"""Generate a macOS app icon for Curl Helper."""
import sys
from PIL import Image, ImageDraw, ImageFont

SIZE = 1024
PAD = int(SIZE * 0.10)  # ~10% margin per Apple HIG
INNER = SIZE - PAD * 2  # actual icon content size
RADIUS = int(INNER * 0.22)
out_path = sys.argv[1] if len(sys.argv) > 1 else "icon_1024.png"

# Draw on inner-sized canvas first
inner_img = Image.new("RGBA", (INNER, INNER), (0, 0, 0, 0))
draw = ImageDraw.Draw(inner_img)

# ── Background gradient ──
for y in range(INNER):
    t = y / INNER
    r = int(18 + (26 - 18) * t)
    g = int(20 + (36 - 20) * t)
    b = int(41 + (66 - 41) * t)
    draw.line([(0, y), (INNER, y)], fill=(r, g, b, 255))

# Rounded rect mask
mask_inner = Image.new("L", (INNER, INNER), 0)
ImageDraw.Draw(mask_inner).rounded_rectangle([(0, 0), (INNER, INNER)], radius=RADIUS, fill=255)
inner_img.putalpha(mask_inner)

# ── Subtle grid lines ──
for y in range(40, INNER, 40):
    draw.line([(0, y), (INNER, y)], fill=(255, 255, 255, 8), width=1)

# ── Accent glow ──
glow = Image.new("RGBA", (INNER, INNER), (0, 0, 0, 0))
glow_draw = ImageDraw.Draw(glow)
for rv in range(200, 0, -2):
    alpha = max(0, int(16 * (1 - rv / 200)))
    cx, cy = INNER // 2, INNER // 2
    glow_draw.ellipse([cx - rv, cy - rv, cx + rv, cy + rv], fill=(230, 113, 84, alpha))
inner_img = Image.alpha_composite(inner_img, glow)
draw = ImageDraw.Draw(inner_img)

# ── Fonts ──
def get_font(size, bold=False):
    names = [
        "/System/Library/Fonts/SFMono-Bold.otf" if bold else "/System/Library/Fonts/SFMono-Regular.otf",
        "/System/Library/Fonts/Menlo.ttc",
        "/System/Library/Fonts/Monaco.dfont",
    ]
    for n in names:
        try:
            return ImageFont.truetype(n, size)
        except (OSError, IOError):
            continue
    return ImageFont.load_default()

def get_light_font(size):
    for n in ["/System/Library/Fonts/SFMono-Light.otf", "/System/Library/Fonts/SFMono-Regular.otf", "/System/Library/Fonts/Menlo.ttc"]:
        try:
            return ImageFont.truetype(n, size)
        except (OSError, IOError):
            continue
    return ImageFont.load_default()

S = INNER  # shorthand
brace_font = get_light_font(int(S * 0.44))
arrow_font = get_font(int(S * 0.23), bold=True)
curl_font = get_font(int(S * 0.10), bold=False)

# ── "{ }" braces ──
teal = (79, 201, 176, 255)
bb = draw.textbbox((0, 0), "{", font=brace_font)
tw = bb[2] - bb[0]
draw.text((S * 0.18 - tw // 2, S * 0.28), "{", fill=teal, font=brace_font)
draw.text((S * 0.78 - tw // 2, S * 0.28), "}", fill=teal, font=brace_font)

# ── ">_" arrow ──
orange = (230, 113, 84, 255)
bb = draw.textbbox((0, 0), ">_", font=arrow_font)
tw = bb[2] - bb[0]
draw.text((S * 0.50 - tw // 2, S * 0.35), ">_", fill=orange, font=arrow_font)

# ── "curl" text ──
light_text = (165, 182, 204, 160)
bb = draw.textbbox((0, 0), "curl", font=curl_font)
tw = bb[2] - bb[0]
draw.text((S * 0.50 - tw // 2, S * 0.78), "curl", fill=light_text, font=curl_font)

# Re-apply mask
masked = Image.new("RGBA", (INNER, INNER), (0, 0, 0, 0))
masked.paste(inner_img, mask=mask_inner)

# ── Place centered on full-size transparent canvas ──
final = Image.new("RGBA", (SIZE, SIZE), (0, 0, 0, 0))
final.paste(masked, (PAD, PAD))
final.save(out_path, "PNG")
print(f"Icon generated: {out_path}")

File: crates/play-gui/curl-helper/scripts/generate_icon.swift
#!/usr/bin/env swift
import Cocoa

let size: CGFloat = 1024
let outPath = CommandLine.arguments.count > 1 ? CommandLine.arguments[1] : "icon_1024.png"

let img = NSImage(size: NSMakeSize(size, size))
img.lockFocus()

guard let ctx = NSGraphicsContext.current?.cgContext else {
    fatalError("No context")
}

// ── Background: macOS-style rounded rect with dark gradient ──
let rect = CGRect(x: 0, y: 0, width: size, height: size)
let radius: CGFloat = size * 0.22
let bgPath = CGMutablePath()
bgPath.addRoundedRect(in: rect, cornerWidth: radius, cornerHeight: radius)
ctx.addPath(bgPath)
ctx.clip()

let cs = CGColorSpaceCreateDeviceRGB()
let bgGrad = CGGradient(colorsSpace: cs, colors: [
    CGColor(srgbRed: 0.07, green: 0.08, blue: 0.16, alpha: 1.0),
    CGColor(srgbRed: 0.10, green: 0.14, blue: 0.26, alpha: 1.0),
] as CFArray, locations: [0.0, 1.0])!
ctx.drawLinearGradient(bgGrad, start: CGPoint(x: 0, y: size), end: CGPoint(x: size, y: 0), options: [])

// ── Subtle grid lines for depth ──
ctx.setStrokeColor(CGColor(srgbRed: 1, green: 1, blue: 1, alpha: 0.03))
ctx.setLineWidth(1.0)
let step: CGFloat = 48
var y: CGFloat = step
while y < size {
    ctx.move(to: CGPoint(x: 0, y: y))
    ctx.addLine(to: CGPoint(x: size, y: y))
    y += step
}
ctx.strokePath()

// ── Helper: draw text centered ──
func drawText(_ text: String, at point: NSPoint, font: NSFont, color: NSColor) {
    let attrs: [NSAttributedString.Key: Any] = [
        .font: font,
        .foregroundColor: color,
    ]
    let str = NSAttributedString(string: text, attributes: attrs)
    let line = CTLineCreateWithAttributedString(str)
    let bounds = CTLineGetBoundsWithOptions(line, .useOpticalBounds)
    ctx.saveGState()
    ctx.textPosition = CGPoint(x: point.x - bounds.width / 2, y: point.y - bounds.height / 2)
    CTLineDraw(line, ctx)
    ctx.restoreGState()
}

// ── Large curly braces "{ }" ──
let braceFont = NSFont.systemFont(ofSize: 420, weight: .thin)
let teal = NSColor(srgbRed: 0.31, green: 0.79, blue: 0.69, alpha: 1.0)

// Left brace
drawText("{", at: NSPoint(x: size * 0.28, y: size * 0.46), font: braceFont, color: teal)
// Right brace
drawText("}", at: NSPoint(x: size * 0.72, y: size * 0.46), font: braceFont, color: teal)

// ── Arrow ">" in the center ── execution symbol
let arrowFont = NSFont.systemFont(ofSize: 200, weight: .bold)
let orange = NSColor(srgbRed: 0.90, green: 0.44, blue: 0.33, alpha: 1.0)
drawText(">", at: NSPoint(x: size * 0.50, y: size * 0.48), font: arrowFont, color: orange)

// ── "curl" text at bottom ──
let curlFont = NSFont.monospacedSystemFont(ofSize: 90, weight: .medium)
let lightGray = NSColor(srgbRed: 0.65, green: 0.72, blue: 0.80, alpha: 0.7)
drawText("curl", at: NSPoint(x: size * 0.50, y: size * 0.14), font: curlFont, color: lightGray)

// ── Accent glow behind arrow ──
ctx.saveGState()
ctx.setBlendMode(.screen)
let glowGrad = CGGradient(colorsSpace: cs, colors: [
    CGColor(srgbRed: 0.90, green: 0.44, blue: 0.33, alpha: 0.15),
    CGColor(srgbRed: 0.90, green: 0.44, blue: 0.33, alpha: 0.0),
] as CFArray, locations: [0.0, 1.0])!
ctx.drawRadialGradient(glowGrad,
    startCenter: CGPoint(x: size * 0.50, y: size * 0.50),
    startRadius: 0,
    endCenter: CGPoint(x: size * 0.50, y: size * 0.50),
    endRadius: size * 0.25,
    options: [])
ctx.restoreGState()

img.unlockFocus()

// ── Save as PNG ──
guard let tiffData = img.tiffRepresentation,
      let rep = NSBitmapImageRep(data: tiffData),
      let png = rep.representation(using: .png, properties: [:]) else {
    fatalError("Failed to create PNG")
}
try! png.write(to: URL(fileURLWithPath: outPath))
print("Icon generated: \(outPath)")
